kotlinx-coroutines-core/common/src/flow/Builders.kt

```diff
@@ -315,3 +324,31 @@ private class ChannelFlowBuilder<T>(
     override fun toString(): String =
         "block[$block] -> ${super.toString()}"
 }
+
+private class CallbackFlowBuilder<T>(
+    private val block: suspend ProducerScope<T>.() -> Unit,
+    context: CoroutineContext = EmptyCoroutineContext,
+    capacity: Int = BUFFERED
+) : ChannelFlowBuilder<T>(block, context, capacity) {
+
+    override suspend fun collectTo(scope: ProducerScope<T>) {
+        super.collectTo(scope)
+        /*
+         * We expect user either call `awaitClose` from within a block (then the channel is closed at this moment)
+         * or being closed/cancelled externally/manually. Otherwise "user forgot to call
+         * awaitClose and receives unhelpful ClosedSendChannelException exceptions" situation is detected.
+         */
+        if (!scope.isClosedForSend) {
+            throw IllegalStateException(
+                """
+                    'awaitClose { yourCallbackOrListener.cancel() }' should be used in the end of callbackFlow block.
+                    Otherwise, a callback/listener may leak in case of cancellation external cancellation (e.g. by 'take(1)' or destroyed activity).
```

Comment:
" ... in case of cancellation external cancellation (e.g. by 'take(1)' or destroyed activity)." -> "cancellation" is repeated here; "destroyed activity" is out of place. I'd actually remove the explanation in the parenthesis altogether. A prompt to read the API docs (where it is all explained in detail) should be enough.